rllib/optimizers/segment_tree.py

```diff
@@ -2,140 +2,149 @@
-
-    def _reduce_helper(self, start, end, node, node_start, node_end):
-        if start == node_start and end == node_end:
-            return self._value[node]
-        mid = (node_start + node_end) // 2
-        if end <= mid:
-            return self._reduce_helper(start, end, 2 * node, node_start, mid)
-        else:
-            if mid + 1 <= start:
-                return self._reduce_helper(start, end, 2 * node + 1, mid + 1,
-                                           node_end)
-            else:
-                return self._operation(
-                    self._reduce_helper(start, mid, 2 * node, node_start, mid),
-                    self._reduce_helper(mid + 1, end, 2 * node + 1, mid + 1,
-                                        node_end))
+            "Capacity must be positive and a power of 2!"
+        self.capacity = capacity
+        if neutral_element is None:
+            neutral_element = 0.0 if operation is operator.add else \
+                float("-inf") if operation is max else float("inf")
+        self.neutral_element = neutral_element
+        self.value = [self.neutral_element for _ in range(2 * capacity)]
+        self.operation = operation
 
     def reduce(self, start=0, end=None):
-        """Returns result of applying `self.operation`
-        to a contiguous subsequence of the array.
+        """Applies `self.operation` to subsequence of our values.
+
+        Subsequence is contiguous, includes `start` and excludes `end`.
 
           self.operation(
               arr[start], operation(arr[start+1], operation(... arr[end])))
 
-        Parameters
-        ----------
-        start: int
-          beginning of the subsequence
-        end: int
-          end of the subsequences
-
-        Returns
-        -------
-        reduced: obj
-          result of reducing self.operation over the specified range of array
-          elements.
+        Args:
+            start (int): Start index to apply reduction to.
+            end (int): Start index to apply reduction to (excluded).
+
+        Returns:
+            any: The result of reducing self.operation over the specified
+                range of `self._value` elements.
         """
         if end is None:
-            end = self._capacity - 1
-        if end < 0:
-            end += self._capacity
-        return self._reduce_helper(start, end, 1, 0, self._capacity - 1)
+            end = self.capacity
+        elif end < 0:
+            end += self.capacity
+
+        # Init result with neutral element.
+        result = self.neutral_element
+        start += self.capacity
+        end += self.capacity
+
+        while start < end:
+            if start & 1:
+                result = self.operation(result, self.value[start])
+                start += 1
+            if end & 1:
+                end -= 1
+                result = self.operation(result, self.value[end])
+            start = start >> 1
+            end = end >> 1
+        return result
```

Comment:
Could you point out which changes were fixing the bug? It's hard to tell since the code was also refactored at the same time.

Reply:
This change here on April 23rd 2018 by RFox.
```
        if end is None:
            end = self._capacity - 1.  # <- used to be `end = self._capacity`
        if end < 0:
            end += self._capacity
        end -= 1  # <- was added
```